### PR TITLE
[ci] Run Linux OpenCL tests against POCL instead of the AMD App SDK

### PR DESCRIPTION
**Pull request by jgiannuzzi**
As discussed in https://github.com/microsoft/LightGBM/pull/5252#discussion_r892772322, I think that using [PoCL](http://portablecl.org) for the Linux `gpu` tests would make a lot of sense, as it is open source and actively maintained, as opposed to the AMD App SDK.

The `dask` tests are currently disabled for the `gpu` task, but they in my local testing the do succeed with PoCL. Should we maybe (re-)enable them?

### CODE REVIEW

**Comment by StrikerRUS** [comment]
> The `dask` tests are currently disabled for the `gpu` task, but they in my local testing the do succeed with PoCL. Should we maybe (re-)enable them?

Here is the reason why they are disabled: https://github.com/microsoft/LightGBM/pull/3708#issuecomment-758202616. I think for this PR we shouldn't enable them to not delay #5252. In a follow-up PR we'll definitely try to re-enable them and run multiple times to ensure their stability.

**Comment by StrikerRUS** [comment]
How about Windows, is PoCL available there?

**Comment by StrikerRUS** [comment]
Could you please update [`GPU Targets Table`](https://github.com/microsoft/LightGBM/blob/master/docs/GPU-Targets.rst#gpu-targets-table) with PoCL in this PR as well?

**Review by StrikerRUS — COMMENTED** [review]
Thank you so much for this PR!

Just some minor comments below.

File: .ci/setup.sh [hunk 1]
```diff
@@ -71,14 +71,21 @@ else  # Linux
...
+                pocl-opencl-icd
+        else
+            sudo apt-get install --no-install-recommends -y \
+                libhwloc-dev \
+                libtinfo-dev \
+                ocl-icd-dev \
+                pkg-config \
+                zlib1g-dev
+            git clone --depth 1 --branch v1.8 https://github.com/pocl/pocl.git
+            cmake -B pocl/build -S pocl -DCMAKE_BUILD_TYPE=release -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS=-stdlib=libc++ -DPOCL_INSTALL_ICD_VENDORDIR=/etc/OpenCL/vendors -DPOCL_DEBUG_MESSAGES=OFF -DSTATIC_LLVM=ON -DINSTALL_OPENCL_HEADERS=OFF -DENABLE_SPIR=OFF -DENABLE_POCLCC=OFF -DENABLE_TESTS=OFF -DENABLE_EXAMPLES=OFF
```
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply .ci/setup.sh:85]: It does require clang/llvm to be installed, but we only build PoCL on Ubuntu 14.04, and our image comes with LLVM 9.0 pre-installed.
PoCL is packaged on more recent distributions, so we don't need to install LLVM explicitly.

File: .vsts-ci.yml [hunk 2]
```diff
@@ -55,21 +55,14 @@ jobs:
       gpu_source:
         TASK: gpu
         METHOD: source
-        # on Ubuntu 14.04, gpu_source build segfaults with newer version of Python
-        # (and newer version of scipy as a consequence)
-        PYTHON_VERSION: '3.7'
+        PYTHON_VERSION: '3.9'
```
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply .vsts-ci.yml:58]: I'll give it a try locally. I set it to Python 3.9 to be consistent with what we do on `ubuntu-latest`.

**Comment by jgiannuzzi** [comment]
> How about Windows, is PoCL available there?

It's available, but unfortunately not as an ICD, which means that we wouldn't be able to test the integrated OpenCL builds, and that the other OpenCL builds would not test something representative.

File: .ci/setup.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply .ci/setup.sh:85]: Ah, indeed, you're right!

We unconditionally (it doesn't matter what [`compiler`](https://github.com/microsoft/LightGBM/blob/11110c540eb84f519da299ea9eb76cc23a0ca570/.vsts-ci.yml#L32) is used at our CI) install Clang in that image:
https://github.com/guolinke/lightgbm-ci-docker/blob/b2d5f990f28631b5dfcb236aa9d40ca271e60577/dockers/ubuntu-14.04/Dockerfile#L45-L62

File: .vsts-ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply .vsts-ci.yml:58]: Thanks!
We try to cover as more environment combinations as possible (combinations of OSes, compilers, Python versions and so on).

**Comment by StrikerRUS** [comment]
> It's available, but unfortunately not as an ICD,

OK, got it! Thank you for the explanation.

File: .vsts-ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply .vsts-ci.yml:58]: It does work locally, committing!

**Comment by jgiannuzzi** [comment]
> Could you please update [`GPU Targets Table`](https://github.com/microsoft/LightGBM/blob/master/docs/GPU-Targets.rst#gpu-targets-table) and [`OpenCL SDK Installation`](https://github.com/microsoft/LightGBM/blob/master/docs/GPU-Windows.rst#opencl-sdk-installation) with PoCL in this PR as well?

I have [updated](https://github.com/microsoft/LightGBM/pull/5282/commits/e146c71fa9a5600d883f47473b359f60e9543ac2) the `GPU Targets Table`. I haven't updated the Windows docs, as I don't think we can recommend using PoCL on that platform.

**Comment by StrikerRUS** [comment]
> I haven't updated the Windows docs, as I don't think we can recommend using PoCL on that platform.

Given this your previous comment https://github.com/microsoft/LightGBM/pull/5282#issuecomment-1153286445, this sounds reasonable. Thank you!

**Comment by jameslamb** [comment]
Love this change, thanks very much for all the hard work @jgiannuzzi !

**Comment by StrikerRUS** [comment]
> The `dask` tests are currently disabled for the `gpu` task, but they in my local testing the do succeed with PoCL. Should we maybe (re-)enable them?

> Here is the reason why they are disabled: https://github.com/microsoft/LightGBM/pull/3708#issuecomment-758202616. I think for this PR we shouldn't enable them to not delay #5252. In a follow-up PR we'll definitely try to re-enable them and run multiple times to ensure their stability.

Unfortunately, with PoCL Dask tests fail with `OSError: dlopen: cannot load any more object with static TLS` error in both 14.04 and 20.04 Ubuntu.
https://github.com/microsoft/LightGBM/pull/5285#issuecomment-1154260141

**Comment by jgiannuzzi** [comment]
> Unfortunately, with PoCL Dask tests fail with `OSError: dlopen: cannot load any more object with static TLS` error in both 14.04 and 20.04 Ubuntu. [#5285 (comment)](https://github.com/microsoft/LightGBM/pull/5285#issuecomment-1154260141)

My understanding is that the `dask` tests only fail with this error on Ubuntu 14.04 with Python 3.10. I ran them locally with Python 3.8 and 3.9 and they work fine. Do we maybe want to set `PYTHON_VERSION` again on the Linux `gpu_source` job?

On Ubuntu 20.04, the tests seem to fail because of a timeout. Enabling those tests on a fake GPU is expected to take more time. Maybe setting `timeoutInMinutes` like we do for the aarch64 job would help?

I'm happy to create a PR for this if you'd like.

**Pull request merged**